### PR TITLE
Customize toolbars

### LINKED ISSUE #12547: Icon size
### Feature description

I am using a high DPI display (267 dpi) and the icons are very small.
I have to move the mouse precise, especially when using the drop down menu of e.g. the stash icon.

A setting that allow to scale the icons to e.g. 150% or 200% would be great.

Note: the "Auto scale user interface when high DPI is used" does not change anything

### Environment

- GitExtensions version: 5.2.1.18601
- GIT version:  2.50.1
- OS version: Win 10.0.19045
- .NET version: 8.0.15

### PR DESCRIPTION
**Pull request by xan2622**
Fixes : #12547

## Proposed changes

This PR lets users fully customize the toolbars of the main window. Instead of three fixed toolbars, users can now create their own, decide which commands appear on each, arrange them on screen, and tune their appearance. Everything is saved and restored between sessions.

### Featuring: 

- Create, edit, rename, and delete your own toolbars, in addition to the three built-in ones.
- Arrange toolbars freely across several rows in a dedicated "_Toolbars Layout_" window, using drag and drop with a live preview.
- Move a toolbar into the empty space between two rows to place it on a brand-new row.
- Set the icon size of each toolbar independently, from small to very large.
- Briefly flash a toolbar in the main window (with the "_Locate_" button) to find it at a glance.
- Choose exactly which commands appear on each toolbar, picked from the full set of main-menu and right-click-menu actions.
- Reorder the commands on a toolbar in the "_Settings - Toolbars_" window by dragging them up or down.
- Place the same command on more than one toolbar at the same time; every copy stays in sync (icon, enabled or disabled state, and checked state).
- Curated command presets to fill a toolbar quickly: "All Actions", the right-click menu, every main-menu category, and the default Standard and Filters sets.
- Show or hide the text label next to icons, both per icon and for a whole toolbar at once.
- A dedicated menu-font setting in the appearance options controls the font of the drop-down menus opened from toolbar buttons.
- Your complete setup (layout, visibility, sizes, and labels) is saved and restored automatically; setups saved by earlier versions are migrated without any action on your part.

### Before

image2

### After

image

image

image

## Test environment(s)

- git version 2.54.0.windows.1
- Windows 11 

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: src/app/GitCommands/Settings/AppSettings.cs [hunk 1]
```diff
@@ -2308,6 +2359,46 @@ public readonly Lazy<string?> ApplicationDataPath
 
         public readonly void ResetDocumentationBaseUrl() => AppSettings._documentationBaseUrl = null;
     }
+
+    [Conditional("DEBUG")]
+    private static void LogToolbarLayout(string message) => Debug.WriteLine(message);
+
+    // Gets or sets the toolbar layout configuration (stored as XML)
+    public static ToolbarLayoutConfig ToolbarLayout
```
Comment (RussKie) [inline src/app/GitCommands/Settings/AppSettings.cs]: IMO, the toolbar configuration is only applicable to the GitUI project, so it should live there. The appsettngs _could_ provide a xml-based settings containing the serialised value, but all the management should not be part of this project.

My gut feel that the toolbar-management related functionality should live somewhere under src/app/GitUI/CommandsDialogs/SettingsDialog/Toolbars folder.

Once moved, most (if not all) these new types can be made non-public (e.g., internal for test purposes)

File: src/app/GitCommands/Settings/AppSettings.cs [hunk 2]
```diff
@@ -1531,6 +1576,12 @@ public static Font Font
         set => SetFont("font", value);
     }
 
+    public static Font MenuFont
+    {
+        get => GetFont("menufont", SystemFonts.MenuFont ?? SystemFonts.MessageBoxFont!);
```
Comment (RussKie) [inline src/app/GitCommands/Settings/AppSettings.cs]: In which situation could `SystemFonts.MenuFont` be null?

File: src/app/GitCommands/Settings/ToolbarLayoutConfig.cs [hunk 3]
```diff
@@ -0,0 +1,185 @@
+﻿using System.Collections.Generic;
+using System.Runtime.Serialization;
+
+namespace GitCommands.Settings;
+
+// Configuration for a single toolbar item (button, menu, etc.)
+[DataContract]
+public class ToolbarItemConfig
```
Comment (RussKie) [inline src/app/GitCommands/Settings/ToolbarLayoutConfig.cs]: One type per file (exception an interface with its implementation).

File: src/app/GitCommands/Settings/ToolbarLayoutConfig.cs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        int row,
+        int orderInRow,
+        bool visible,
+        int iconSize,
+        int? index = null,
+        bool allIconsShowText = false)
+    {
+        // --- ToolbarsVisibility ---
+        ToolbarMetadata? visMeta = ToolbarsVisibility.FirstOrDefault(t => t.Name == name);
+        if (visMeta != null)
```
Comment (RussKie) [inline src/app/GitCommands/Settings/ToolbarLayoutConfig.cs]: `is/is not null`

File: src/app/GitCommands/Settings/ToolbarLayoutConfig.cs [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    public int IconSize { get; set; } = 16;
+
+    // Whether all icons in this toolbar show their text label ("For all icons" mode).
+    [DataMember]
+    public bool AllIconsShowText { get; set; }
+}
+
+// Metadata for a built-in or custom toolbar
+[DataContract]
+public class ToolbarMetadata
```
Comment (RussKie) [inline src/app/GitCommands/Settings/ToolbarLayoutConfig.cs]: `BuiltInToolbarMetadata`
(though I'd prefer `ToolbarBuiltInMetadata` and `ToolbarCustomMetadata` purely from the file/type-grouping perspective)

File: src/app/GitCommands/Settings/ToolbarLayoutConfig.cs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+{
+    // Name of the toolbar (e.g., "Standard", "Filters", "Scripts", "Custom 01")
+    [DataMember]
+    public string Name { get; set; } = string.Empty;
```
Comment (RussKie) [inline src/app/GitCommands/Settings/ToolbarLayoutConfig.cs]: There's no guarantee or enforcement of this being a valid string (besides not being null or empty).

File: src/app/GitUI/CommandsDialogs/IPushLabelItem.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+/// Implemented by both <see cref="ToolStripPushButton"/> (the original on the Standard toolbar)
+/// and <see cref="ToolStripPushButtonClone"/> (independent copies on custom toolbars).
+/// Allows <see cref="FormBrowse"/> infrastructure to treat both uniformly without resolving
+/// the original from a clone's Tag.
+/// </summary>
+internal interface IPushLabelItem
+{
+    bool ShowLabel { get; set; }
+
+    string? LabelText { get; set; }
```
Comment (RussKie) [inline src/app/GitUI/CommandsDialogs/IPushLabelItem.cs:11]: A-Z order please

File: src/app/GitUI/Resources/Icons/arrow-090.png
Comment (RussKie) [inline src/app/GitUI/Resources/Icons/arrow-090.png:1]: What's the origin of the images? Did they come from https://p.yusukekamiyamane.com/?

**Review by RussKie — COMMENTED** [review]
I haven't run, simply looked at the provided screenshots - this is very cool.

File: src/app/GitUI/Resources/Icons/arrow-090.png
Comment (xan2622) [reply src/app/GitUI/Resources/Icons/arrow-090.png:1]: Yes, they do.

File: src/app/GitUI/CommandsDialogs/SettingsDialog/Toolbars/ToolbarLayoutConfig.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    /// available index is computed automatically from the current <see cref="CustomToolbars"/> count.
+    /// </param>
+    /// <param name="allIconsShowText">Whether all icons show their text label.</param>
+    public void SetCustomToolbarMetadata(
+        string name,
+        int row,
+        int orderInRow,
+        bool visible,
+        int iconSize,
+        int? index = null,
```
Comment (RussKie) [inline src/app/GitUI/CommandsDialogs/SettingsDialog/Toolbars/ToolbarLayoutConfig.cs:44]: All of the data must be validated - e.g., valid names, valid order (0+ - some valid range), valid icon sizes (16x16, 24x24, 48x48, etc., with some valid upper limit), etc.

File: src/app/GitUI/CommandsDialogs/SettingsDialog/Toolbars/ToolbarLayoutStore.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+// Loads and saves the strongly-typed toolbar layout configuration, backed by the raw XML
+// stored in AppSettings.ToolbarLayoutXml. Toolbar customization is only meaningful to GitUI,
+// so the typed model and its (de)serialization live here rather than in GitCommands.
+internal static class ToolbarLayoutStore
+{
+    // Never returns null: an absent or unreadable setting yields an empty configuration.
+    public static ToolbarLayoutConfig Load()
+    {
+        return ToolbarXmlSerializer.Deserialize<ToolbarLayoutConfig>(AppSettings.ToolbarLayoutXml) ?? new ToolbarLayoutConfig();
```
Comment (RussKie) [inline src/app/GitUI/CommandsDialogs/SettingsDialog/Toolbars/ToolbarLayoutStore.cs]: All deserialised data must be validated

**Review by RussKie — COMMENTED** [review]
Please add tests for the new and changed functionality

File: src/app/GitUI/CommandsDialogs/SettingsDialog/Toolbars/ToolbarXmlSerializer.cs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    public static T? Deserialize<T>(string xml) where T : class
+    {
+        if (string.IsNullOrWhiteSpace(xml))
+        {
+            return null;
+        }
+
+        try
+        {
+            DataContractSerializer serializer = new(typeof(T));
```
Comment (RussKie) [inline src/app/GitUI/CommandsDialogs/SettingsDialog/Toolbars/ToolbarXmlSerializer.cs]: **IgorBot (AI review): bound both the XML and the resulting object graph before deserializing.** `DataContractSerializer(Type)` defaults `MaxItemsInObjectGraph` to `int.MaxValue`, and `ReadObject(Stream)` creates a reader with `XmlDictionaryReaderQuotas.Max`. `Encoding.UTF8.GetBytes(xml)` also duplicates the entire unbounded setting before any parser limit applies. A corrupted or poisoned settings value can therefore exhaust memory/CPU before this catch block can recover. Please reject oversized input up front, set a domain-sized `MaxItemsInObjectGraph`, and deserialize through a reader with explicit depth/string/document limits (with DTD processing prohibited and no resolver). Catch expected parse/serialization exceptions rather than `Exception`; recovering from `OutOfMemoryException` is not reliable.

File: src/app/GitUI/CommandsDialogs/SettingsDialog/Toolbars/ToolbarLayoutStore.cs (hunk 9 above)
Comment (RussKie) [inline src/app/GitUI/CommandsDialogs/SettingsDialog/Toolbars/ToolbarLayoutStore.cs]: **IgorBot (AI review): make this the single atomic validation boundary before returning a config.** `DataContractSerializer` can populate `null` members (including via `xsi:nil`, and constructors/field initializers are not invariants to rely on), so the non-null annotations do not make `Items`, `CustomToolbars`, `ToolbarsVisibility`, or their string fields trustworthy. Validate or normalize into a separate domain object before any caller sees it, otherwise values such as a huge `IconSize` flow into `Size`/`Font` construction during startup. At minimum enforce collection and string length caps; non-null entries; unique toolbar names and custom indices; the exact icon-size allowlist; bounded non-negative row/order/index values; built-in/custom name separation; item-to-toolbar referential integrity; and the allowed item-ID/label placeholder grammar. Reject the whole payload to the empty default on any violation, and add adversarial tests for missing/nil members, duplicates, extreme integers, oversized collections/strings, malformed labels, and unknown item IDs.

**Review by RussKie — COMMENTED** [review]
IgorBot (AI-assisted security review): I found no apparent arbitrary-type deserialization or automatic command-execution path here: DataContractSerializer is given a fixed root type and no custom resolver/known-type expansion. The actionable risks are availability and invalid state crossing the settings boundary. The inline comments recommend bounded parser/object-graph quotas and centralized semantic validation.
